sql: use the new EXPLAIN infrastructure for UI plans

RaduBerinde opened the pull request:
#### sql: don't key statement statistics on distribution/vectorization

The statement statistics code buckets statements separately depending on the
distribution and vectorization flags. This is not very useful: we normally don't
expect these flags to change for the same statement; and the UI doesn't show
them separately anyway. In addition, it poses a problem to switching the explain
tree (presented to UI) to use the new explain infrastructure: we need to know if
we will need the tree upfront (before execbuild).

This change removes these flags from the key. The flags are still populated in
the `StatementStatisticsKey`.

Release note: None

#### sql: use the new EXPLAIN infrastructure for UI plans

This change reworks how the `roachpb.ExplainPlanTreeNode` tree (used by the UI
to show statement plans) is populated. We use the new EXPLAIN infrastructure.
The main change is that we need to decide upfront if we will need the plan or
not.

Release note (admin ui change): various improvements to the statement plans in
the UI.

RaduBerinde committed:
sql: use the new EXPLAIN infrastructure for UI plans

This change reworks how the `roachpb.ExplainPlanTreeNode` tree (used by the UI
to show statement plans) is populated. We use the new EXPLAIN infrastructure.
The main change is that we need to decide upfront if we will need the plan or
not.

Release note (admin ui change): various improvements to the statement plans in
the UI.

RaduBerinde committed:
sql: don't key statement statistics on distribution/vectorization

The statement statistics code buckets statements separately depending on the
distribution and vectorization flags. This is not very useful: we normally don't
expect these flags to change for the same statement; and the UI doesn't show
them separately anyway. In addition, it poses a problem to switching the explain
tree (presented to UI) to use the new explain infrastructure: we need to know if
we will need the tree upfront (before execbuild).

This change removes these flags from the key. The flags are still populated in
the `StatementStatisticsKey`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52956)
<!-- Reviewable:end -->

RaduBerinde committed:
sql: pass exec factory to apply-join/recursive-cte functions

To support executing plans that were generated with `explain.Factory`, we must
not use the same factory for "inner" plans generated by apply joins and
recursive CTEs. We modify the functions to pass the exec factory. Note that we
were already doing this for cascade planning.

Release note: None

RaduBerinde commented:
TFTR!

RaduBerinde commented:
bors r+
